js/ui/grid_core/ui.grid_core.header_panel.js

```diff
@@ -116,10 +152,28 @@ const HeaderPanel = ColumnsView.inherit({
     },
 
     optionChanged: function(args) {
+        const parts = getPathParts(args.fullName);
+
         if(args.name === 'onToolbarPreparing') {
             this._invalidate();
             args.handled = true;
         }
+        if(args.name === 'toolbar') {
+            if(parts.length <= 2) {
+                // toolbar and toolbar.items case
+                const toolbarOptions = this._getToolbarOptions();
+                this._toolbar.option(toolbarOptions);
+            } else if(parts.length === 3) {
+                // toolbar.items[i] case
+                const normalizedItem = this._normalizeToolbarItems(this._getToolbarItems(), args.value);
+                const optionName = args.fullName.replace(/^toolbar./, '');
```

Comment:
The optionName calculation is  duplicated in two cases (parts.length === 3 and parts.length >= 4). May be combine this two cases in one condition? 
```javascript
if(parts.length >= 3) {
    const value = parts.length === 3 ? this._normalizeToolbarItems(this._getToolbarItems(), args.value) : args.value;
    const optionName = args.fullName.replace(/^toolbar./, '');
    this._toolbar.option(optionName, value);
}
```